src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -216,7 +216,30 @@ StorageReplicatedMergeTree::StorageReplicatedMergeTree(
         getStorageID().getFullTableName() + " (StorageReplicatedMergeTree::mutationsFinalizingTask)", [this] { mutationsFinalizingTask(); });
 
     if (global_context.hasZooKeeper())
-        current_zookeeper = global_context.getZooKeeper();
+    {
+        /// It's possible for getZooKeeper() to timeout if  zookeeper host(s) can't
+        /// be reached. In such cases Poco::Exception is thrown after a connection
+        /// timeout - refer to src/Common/ZooKeeper/ZooKeeperImpl.cpp:866 for more info.
+        ///
+        /// Side effect of this is that the CreateQuery gets interrupted and it exits.
+        /// But the data Directories for the tables being created aren't cleaned up.
+        /// This unclean state will hinder table creation on any retries and will
+        /// complain that the Directory for table already exists.
+        ///
+        /// To acheive a clean state on failed table creations, catch this error if
+        /// the excaption is of type Poco::Exception and call dropIfEmpty() method,
+        /// then proceed throwing the exception. Without this, the Directory for the
+        /// tables need to be manually deleted before retrying the CreateQuery.
+        try
+        {
+            current_zookeeper = global_context.getZooKeeper();
+        }
+        catch (Poco::Exception & e)
+        {
+            dropIfEmpty();
```

Comment:
So folders are created in MergeTreeData constructor?

I don't think we should do any destructive operations (even removing empty folders) during ATTACH. 
```cpp
 if (!attach) .... 
```

it seem you  don't need to capture the exception in the variable, so you can do this way:

https://github.com/ClickHouse/ClickHouse/blob/dee1fefeb472bebdfa4e15eb1b7ec7e428069f8a/src/Storages/StorageReplicatedMergeTree.cpp#L321-L325

Reply:
Thanks for suugestions

Reply:
Thanks you. Also just made minor changes to the comment to reflect that this is what we're doing.